### DISCUSSION TITLE
How can I export the screen frames, either as images or gif?

### DISCUSSION THREAD

**Discussion by darrenparkinson** [hajimehoshi/ebiten#2492] [Q&A]
I've just tried ebitengine for an AoC visualisation.  I was wondering how I might easily save the resulting visualisation as a gif?

I tried the EBITENGINE_INTERNAL_IMAGES_KEY and also the EBITENGINE_SCREESHOT_KEY, but not sure they are exactly what I'm after (although I can't quite figure out what the EBITENGINE_INTERNAL_IMAGES_KEY is for 😬)

Overall there would be about 240 frames from start to finish.  I could advance each frame by pressing a key followed by the screenshot key, but it feels like there is probably a better solution.

Any thoughts or pointers in the right direction greatly appreciated.

Many thanks.

**Comment by hajimehoshi** [comment]
There is no easy way to output a GIF animation. I would do

* Convert the rendered screen image (the argument of your `Draw` function) to a palleted image by [FloydSteinberg](https://pkg.go.dev/image/draw)
* Keep the converted image to a [`gif.GIF`](https://pkg.go.dev/image/gif#GIF) object (you need 240 palletted images for 240 frames, but you might be able to reduce the number by adjusting the time interval)
* Dump the GIF object as a gif image

If the quality is not good, [APNG](https://github.com/kettek/apng) is another options.

**Reply by darrenparkinson** [reply]
Thanks for the reply.  I wondered if I might do something like that, but wasn't quite sure how to do it.  Also, `screen` is an `*ebiten.Image` not an `image.Image` which I felt might be an issue? Apologies for my lack of knowledge on this subject, it seems I have a lot of research to do around image formats etc.  Thanks again.

**Reply by hajimehoshi** [reply]
*ebiten.Image is an image.Image. So you can pass an *ebiten.Image to APIs as image.Image.

**Reply by darrenparkinson** [reply]
Ah ok perfect! Thanks.

**Comment by SolarLune** [comment]
Easiest way to do it if you don't need perfect quality is just to record the screen with OBS or something similar, and then convert it to a GIF using something like GifCurry (on Linux), or use a screen recorder.

EDIT: Oh, or yeah, just render images for each frame and create a GIF yourself. This process would actually be fairly easy to package up if you wanted to make it accessible to others, as well.

**Reply by darrenparkinson** [reply]
Hi, thanks, yes, this is the route I've taken so far. I was just wondering if I could build it in to make it a bit easier... I'll play around with it some more. Thanks again.
